src/node/interfaces.cpp

```diff
@@ -723,6 +723,11 @@ class ChainImpl : public Chain
             notifications.transactionAddedToMempool(entry.GetSharedTx(), 0 /* mempool_sequence */);
         }
     }
+    bool hasAssumedValidChain() override
+    {
+        return Assert(m_node.chainman)->IsSnapshotActive();
+    }
```

Comment:
This can be `private` rather than `public`.

Reply:
See its `private` neighbor, `ChainstateManager& chainman() { return *Assert(m_node.chainman); }`. It looks like `hasAssumedValidChain()` may not only be private but also simplified, if wanted, to use `chainman()`. Clang 15 doesn't mind.
```diff
@@ -502,6 +502,10 @@ class ChainImpl : public Chain
 {
 private:
     ChainstateManager& chainman() { return *Assert(m_node.chainman); }
+    bool hasAssumedValidChain() override
+    {
+        return chainman().IsSnapshotActive();
+    }
 public:
     explicit ChainImpl(NodeContext& node) : m_node(node) {}
@@ -767,10 +771,6 @@ public:
             notifications.transactionAddedToMempool(entry.GetSharedTx(), 0 /* mempool_sequence */);
         }
     }
-    bool hasAssumedValidChain() override
-    {
-        return Assert(m_node.chainman)->IsSnapshotActive();
-    }
```

Reply:
I agree with jamesob. Just because a diff compiles, doesn't mean it makes sense.

Reply:
`hasAssumedValidChain` is called by `CWallet::AttachChain()`, which is itself private

Reply:
>  hasAssumedValidChain is called by CWallet::AttachChain(), which is itself private


`private` is used to mark a symbol "private" to a class. For example, `AttachChain` can be private to `CWallet`, meaning only code in the scope of `CWallet` (aka member functions) can call `AttachChain`.

If a function is called from outside its class, it should not be marked `private`. Even if adding `private` happens to compile due to language design trade-offs.

Reply:
In the virtual base class, no. In the derived child the implementation method can be private for both clang and gcc; why?

-----

Anyway, as a reference to refer to if useful here, 693414d2718 added+used the `Chainman()` accessor in the `ChainImpl` class.

```diff
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -769,7 +769,7 @@ public:
     }
     bool hasAssumedValidChain() override
     {
-        return Assert(m_node.chainman)->IsSnapshotActive();
+        return chainman().IsSnapshotActive();
     }
```

Reply:
> In the virtual base class, no. In the derived child the implementation method can be private for both clang and gcc; why?

Found some answers to this. Separation of interface from implementation, and virtual functions exist to allow customization; unless they also need to be invoked directly from within derived classes' code, there's no need to ever make them anything but private.

- Herb Sutter in http://www.gotw.ca/publications/mill18.htm: "Make virtual functions private, and only if derived classes need to invoke the base implementation of a virtual function, make the virtual function protected. Prefer to make interfaces non-virtual using the Template pattern."

- https://en.cppreference.com/w/cpp/language/virtual: "Base::vf does not need to be accessible or visible to be overridden. (Base::vf can be declared private, or Base can be inherited using private inheritance)"

- "A (virtual) implementation member should be protected if it provides an operation or data that a derived class will need to use in its own implementation. Otherwise, implementation members should be private." - C++ Primer (Lippman, Lajoie, Moo)

Reply:
The page from Canada doesn't load for me and making the virtual function private doesn't compile for me (see above). If making the overridden function private is a goal, I'd say it should be done for all of them for consistency, and not only for the one added here.

Reply:
Interesting, it doesn't load on my phone but from my computer with Firefox, it does. Maybe due to not being https.

Yes, the child builds when private but not the base vf, due to the design I guess. I'm going to re-review my changes in #24150 in light of what I'm learning.